### ISSUE TITLE
Repository not found

### ISSUE THREAD

**Issue by ianyoung** [google-github-actions/deploy-cloudrun#420]
### TL;DR

After tagging the docker image it seems to get stuck on the pushing of the image and ends in a "Repository 'NAME' not found" error. 

### Action YAML

```yaml
name: Build and Deploy to Cloud Run

on:
  push:
    branches: [ "main" ]

env:
  PROJECT_ID: my-project # Google Cloud project id
  GAR_LOCATION: europe-west2 # London - Artifact Registry location
  SERVICE: my-service # Cloud Run service name
  REGION: europe-west6 # Cloud Run service region

jobs:
  deploy:
    # Add 'id-token' with the intended permissions for workload identity federation
    permissions:
      contents: 'read'
      id-token: 'write'

    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Google Auth
        id: auth
        uses: 'google-github-actions/auth@v0'
        with:
          token_format: 'access_token'
          workload_identity_provider: '${{ secrets.WIF_PROVIDER }}'
          service_account: '${{ secrets.WIF_SERVICE_ACCOUNT }}'

      # BEGIN - Docker auth and build (NOTE: If you already have a container image, these Docker steps can be omitted)

      # Authenticate Docker to Google Cloud Artifact Registry
      - name: Docker Auth
        id: docker-auth
        uses: 'docker/login-action@v1'
        with:
          username: 'oauth2accesstoken'
          password: '${{ steps.auth.outputs.access_token }}'
          registry: '${{ env.GAR_LOCATION }}-docker.pkg.dev'

      - name: Build and Push Container
        run: |-
          docker build -t "${{ env.GAR_LOCATION }}-docker.pkg.dev/${{ env.PROJECT_ID }}/${{ env.SERVICE }}:${{ github.sha }}" ./src/
          docker push "${{ env.GAR_LOCATION }}-docker.pkg.dev/${{ env.PROJECT_ID }}/${{ env.SERVICE }}:${{ github.sha }}"

      # END - Docker auth and build

      - name: Deploy to Cloud Run
        id: deploy
        uses: google-github-actions/deploy-cloudrun@v0
        with:
          service: ${{ env.SERVICE }}
          region: ${{ env.REGION }}
          # NOTE: If using a pre-built image, update the image name here
          image: ${{ env.GAR_LOCATION }}-docker.pkg.dev/${{ env.PROJECT_ID }}/${{ env.SERVICE }}:${{ github.sha }}

      # If required, use the Cloud Run url output in later steps
      - name: Show Output
        run: echo ${{ steps.deploy.outputs.url }}
```

### Log output

```text
Step 21/21 : CMD ["uvicorn", "app.main:app", "--workers", "1", "--host", "0.0.0.0", "--port", "8000", "--log-config", "config/log_conf.yml"]
 ---> Running in 37ea26d799f8
Removing intermediate container 37ea26d799f8
 ---> d29692719ea5
Successfully built d29692719ea5
Successfully tagged europe-west2-docker.pkg.dev/***/MY-SERVICE:36ff709d0c5e38299da014336b4e9f280a0140a6
The push refers to repository [europe-west2-docker.pkg.dev/***/MY-SERVICE]
dba7b3efb489: Preparing
ec3c4788435a: Waiting
name unknown: Repository "MY-SERVICE" not found
Error: Process completed with exit code 1.
```

### Additional information

My `docker build -t` command includes and addition `./src/` as the Dockerfile sits in there.

I did not create a repository in Artifact Registry beforehand.

When I look in Artifact Registry in the console I see my repository but the path is strange:    
`europe-west2-docker.pkg.dev/MY-PROJECT/MY-PROJECT`

I'm not sure why it seems to have a nested project folder. This repository is also empty.

I've used the exact template as provided with the GitHub Action.

**Comment by sethvargo** [comment]
Hi there - could you please provide the [debug output](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging) for the complete GitHub Actions workflow run?

If you are only building the container to deploy to Cloud Run, you may want to use the `source` option instead of the `image` option.

You also need to make sure that the deploying service account has permissions to read from the Artifact Registry.

**Comment by ianyoung** [comment]
Hi @sethvargo. Thanks for the quick response. 

I've managed to find the issue and get things working but it does indicate an issue with the template provided.

In [troubleshooting advice for Artifact Registry](https://cloud.google.com/artifact-registry/docs/docker/troubleshoot#push) they mention:

> Verify that the full path of the image that you are pushing is correct. The path must include the registry hostname, Google Cloud project ID, **repository**, and image. For example:
```
us-east1-docker.pkg.dev/my-project/my-repo/my-image:v1
```
> Each Artifact Registry repository is a separate resource, so **you cannot push an image to a path without a repository**. For example, us-east1-docker.pkg.dev/my-project/my-image:v1 is an invalid image path.

The `build` and `push` commands from the template however seem to miss out the repository:
```
docker build -t "${{ env.GAR_LOCATION }}-docker.pkg.dev/${{ env.PROJECT_ID }}/${{ env.SERVICE }}:${{ github.sha }}" ./src/
docker push "${{ env.GAR_LOCATION }}-docker.pkg.dev/${{ env.PROJECT_ID }}/${{ env.SERVICE }}:${{ github.sha }}"
```
The same with the `deploy` step:
```
image: ${{ env.GAR_LOCATION }}-docker.pkg.dev/${{ env.PROJECT_ID }}/${{ env.SERVICE }}:${{ github.sha }}
```

I tried changing those to include a hard-coded `MY-REPO` value and that worked.
```
${{ env.GAR_LOCATION }}-docker.pkg.dev/${{ env.PROJECT_ID }}/MY-REPO/${{ env.SERVICE }}:${{ github.sha }}
```

That then matches up with the nested folders I see in Artifact Registry via the console:
`europe-west2-docker.pkg.dev/MY-PROJECT/MY-REPO/MY-SERVICE`

Both the `push` and the follow-up `deploy` work with this change.

Debugging logs are [here](https://d.pr/f/5gStSs) if you still need them.

> If you are only building the container to deploy to Cloud Run, you may want to use the source option instead of the image option.

I thought the `source` option was if you didn't provide a `Dockerfile`? In my case I provide a very specific Dockerfile and I want the control over how it is built. 

I am only deploying to Cloud Run but I do build and work with the container locally in development.

> You also need to make sure that the deploying service account has permissions to read from the Artifact Registry.

All permissions granted. Tested with the above edit and working without any further modifications.

**Comment by sethvargo** [comment]
Hi @ianyoung - I'm glad you were able to get unstuck.

> with the template provided.

Sorry - which template? I don't think there's a template in this repository. If you point me to the template, I can try and get it updated.

> I thought the source option was if you didn't provide a Dockerfile? In my case I provide a very specific Dockerfile and I want the control over how it is built.

`source` will use a Dockerfile if present, otherwise it uses a detected buildpack.

> All permissions granted. Tested with the above edit and working without any further modifications.

Sorry, that was a red herring. I missed the repo format.

**Comment by ianyoung** [comment]
Hmm. I got a pre-populated template when searching through the GitHub Actions for "Deploy Cloud Run":

<img width="1017" alt="github-actions-workflows" src="https://user-images.githubusercontent.com/66717/209693487-f0049c9e-9f17-49b4-8b69-c5c185d36061.png">

Clicking on "Configure" for Build and Deploy to Cloud Run gave a prepopulated yaml file:

```yaml
# This workflow build and push a Docker container to Google Artifact Registry and deploy it on Cloud Run when a commit is pushed to the "main" branch
#
# Overview:
#
# 1. Authenticate to Google Cloud
# 2. Authenticate Docker to Artifact Registry
# 3. Build a docker container
# 4. Publish it to Google Artifact Registry
# 5. Deploy it to Cloud Run
#
# To configure this workflow:
#
# 1. Ensure the required Google Cloud APIs are enabled:
#
#    Cloud Run            run.googleapis.com
#    Artifact Registry    artifactregistry.googleapis.com
#
# 2. Create and configure Workload Identity Federation for GitHub (https://github.com/google-github-actions/auth#setting-up-workload-identity-federation)
#
# 3. Ensure the required IAM permissions are granted
#
#    Cloud Run
#      roles/run.admin
#      roles/iam.serviceAccountUser     (to act as the Cloud Run runtime service account)
#
#    Artifact Registry
#      roles/artifactregistry.admin     (project or repository level)
#
#    NOTE: You should always follow the principle of least privilege when assigning IAM roles
#
# 4. Create GitHub secrets for WIF_PROVIDER and WIF_SERVICE_ACCOUNT
#
# 5. Change the values for the GAR_LOCATION, SERVICE and REGION environment variables (below).
#
# NOTE: To use Google Container Registry instead, replace ${{ env.GAR_LOCATION }}-docker.pkg.dev with gcr.io
#
# For more support on how to run this workflow, please visit https://github.com/marketplace/actions/deploy-to-cloud-run
#
# Further reading:
#   Cloud Run IAM permissions                 - https://cloud.google.com/run/docs/deploying
#   Artifact Registry IAM permissions         - https://cloud.google.com/artifact-registry/docs/access-control#roles
#   Container Registry vs Artifact Registry   - https://cloud.google.com/blog/products/application-development/understanding-artifact-registry-vs-container-registry
#   Principle of least privilege              - https://cloud.google.com/blog/products/identity-security/dont-get-pwned-practicing-the-principle-of-least-privilege

name: Build and Deploy to Cloud Run

on:
  push:
    branches: [ "main" ]

env:
  PROJECT_ID: YOUR_PROJECT_ID # TODO: update Google Cloud project id
  GAR_LOCATION: YOUR_GAR_LOCATION # TODO: update Artifact Registry location
  SERVICE: YOUR_SERVICE_NAME # TODO: update Cloud Run service name
  REGION: YOUR_SERVICE_REGION # TODO: update Cloud Run service region

jobs:
  deploy:
    # Add 'id-token' with the intended permissions for workload identity federation
    permissions:
      contents: 'read'
      id-token: 'write'

    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Google Auth
        id: auth
        uses: 'google-github-actions/auth@v0'
        with:
          token_format: 'access_token'
          workload_identity_provider: '${{ secrets.WIF_PROVIDER }}' # e.g. - projects/123456789/locations/global/workloadIdentityPools/my-pool/providers/my-provider
          service_account: '${{ secrets.WIF_SERVICE_ACCOUNT }}' # e.g. - my-service-account@my-project.iam.gserviceaccount.com

      # NOTE: Alternative option - authentication via credentials json
      # - name: Google Auth
      #   id: auth
      #   uses: 'google-github-actions/auth@v0'
      #   with:
      #     credentials_json: '${{ secrets.GCP_CREDENTIALS }}''

      # BEGIN - Docker auth and build (NOTE: If you already have a container image, these Docker steps can be omitted)

      # Authenticate Docker to Google Cloud Artifact Registry
      - name: Docker Auth
        id: docker-auth
        uses: 'docker/login-action@v1'
        with:
          username: 'oauth2accesstoken'
          password: '${{ steps.auth.outputs.access_token }}'
          registry: '${{ env.GAR_LOCATION }}-docker.pkg.dev'

      - name: Build and Push Container
        run: |-
          docker build -t "${{ env.GAR_LOCATION }}-docker.pkg.dev/${{ env.PROJECT_ID }}/${{ env.SERVICE }}:${{ github.sha }}" ./
          docker push "${{ env.GAR_LOCATION }}-docker.pkg.dev/${{ env.PROJECT_ID }}/${{ env.SERVICE }}:${{ github.sha }}"

      # END - Docker auth and build

      - name: Deploy to Cloud Run
        id: deploy
        uses: google-github-actions/deploy-cloudrun@v0
        with:
          service: ${{ env.SERVICE }}
          region: ${{ env.REGION }}
          # NOTE: If using a pre-built image, update the image name here
          image: ${{ env.GAR_LOCATION }}-docker.pkg.dev/${{ env.PROJECT_ID }}/${{ env.SERVICE }}:${{ github.sha }}

      # If required, use the Cloud Run url output in later steps
      - name: Show Output
        run: echo ${{ steps.deploy.outputs.url }}
```

So is there any benefit in using the Build from Source option? The only difference is that it seems to remove the building and pushing of the container to Artifact Registry.

**Comment by sethvargo** [comment]
@verbanicm can you look at the starter workflow? Looks like it might have a bug.

@ianyoung the build from source option will reduce the amount of code in your workflow and Google Cloud will manage the container images over time. If you build yourself, you're responsible for cleaning up yourself too.

**Comment by ianyoung** [comment]
Ok, that does sound useful. I was wondering if there was an automated way to clean up old images. I'll give the build from source option a go.

Thanks.
